src/core/ddsi/src/ddsi_typelib.c

```diff
@@ -919,4 +1123,70 @@ void ddsi_type_pair_free (struct ddsi_type_pair *type_pair)
   ddsrt_free (type_pair);
 }
 
+dds_return_t ddsi_wait_for_type_resolved (struct ddsi_domaingv *gv, const ddsi_typeid_t *type_id, dds_duration_t timeout, struct ddsi_type **type, ddsi_type_include_deps_t resolved_kind, ddsi_type_request_t request)
+{
+  dds_return_t ret;
+
+  assert (type);
+  if (ddsi_typeid_is_none (type_id) || !ddsi_typeid_is_hash (type_id))
+  {
+    ret = DDS_RETCODE_BAD_PARAMETER;
+    goto err;
+  }
+
+  ddsrt_mutex_lock (&gv->typelib_lock);
+  *type = ddsi_type_lookup_locked (gv, type_id);
+  if (*type == NULL)
+  {
+    /* For a type to be resolved, we require it's top-level type identifier to be known
+       and added to the type library as a result of a discovered endpoint or topic,
+       or a topic created locally. */
+    ret = DDS_RETCODE_PRECONDITION_NOT_MET;
+    goto err_unlock;
+  }
+
+  if (ddsi_type_resolved_locked (gv, *type, resolved_kind))
+  {
+    ddsi_type_ref_locked (gv, NULL, *type);
+    ret = DDS_RETCODE_OK;
+    goto resolved;
+  }
+  else if (timeout == 0)
+  {
+    ret = DDS_RETCODE_TIMEOUT;
+    goto err_unlock;
+  }
+  ddsrt_mutex_unlock (&gv->typelib_lock);
+
+  if (request == DDSI_TYPE_SEND_REQUEST)
+  {
+    // FIXME: provide proxy pp guid
+    if (!ddsi_tl_request_type (gv, type_id, NULL, resolved_kind))
+    {
+      ret = DDS_RETCODE_PRECONDITION_NOT_MET;
+      goto err;
+    }
+  }
+
+  const dds_time_t tnow = dds_time ();
+  const dds_time_t abstimeout = (DDS_INFINITY - timeout <= tnow) ? DDS_NEVER : (tnow + timeout);
+  ddsrt_mutex_lock (&gv->typelib_lock);
+  while (!ddsi_type_resolved_locked (gv, *type, resolved_kind))
+  {
+    if (!ddsrt_cond_waituntil (&gv->typelib_resolved_cond, &gv->typelib_lock, abstimeout))
+    {
+      ret = DDS_RETCODE_TIMEOUT;
+      goto err_unlock;
+    }
+  }
+  ddsi_type_ref_locked (gv, NULL, *type);
+  ret = DDS_RETCODE_OK;
+
+resolved:
```

Comment:
Rather than `resolved` and `err_unlock`, I'd probably go for `out`. Or, more likely even, try to refactor it into multiple functions so that a simple `return` statement sufficed in the prepatory parts in an inner function, with an outer function that just waits and locks/unlocks the mutex.

Reply:
Good point, the refactoring is in [9eb81bc](https://github.com/eclipse-cyclonedds/cyclonedds/pull/1308/commits/9eb81bc832506d2dac3297ddb603fffe440d7879)